File: IPWorks Samples/WebDAV Client/c++/webdav.cpp
/*
 * IPWorks 2024 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string>
#include <string.h>
#include <stdlib.h>
#include <time.h>
#include <vector>
#include "../../include/ipworks.h"

using namespace std;

#define LINE_LEN 256

class MyWebDAV : public WebDAV
{
public:
	bool exportingEvent;
	int eventNumber;
	vector<string> resourceURIs;

	MyWebDAV()
	{
		eventNumber = 1;
		exportingEvent = false;
	}

	virtual int FireSSLServerAuthentication(WebDAVSSLServerAuthenticationEventParams* e)
	{
		if (e->Accept) return 0;
		printf("Server provided the following certificate:\nIssuer: %s\nSubject: %s\n",
			e->CertIssuer, e->CertSubject);
		printf("The following problems have been determined for this certificate: %s\n", e->Status);
		printf("Would you like to continue anyways? [y/n] ");
		if (getchar() == 'y') e->Accept = true;
		else exit(0);
		return 0;
	}
	virtual int FireConnected(WebDAVConnectedEventParams* e)
	{
		printf("Server connected\n");
		return 0;
	}
	virtual int FireConnectionStatus(WebDAVConnectionStatusEventParams* e)
	{
		printf("Status code %i: %s\n", e->StatusCode, e->Description);
		return 0;
	}
	virtual int FireDisconnected(WebDAVDisconnectedEventParams* e)
	{
		printf("Server Disconnected\n");
		return 0;
	}
	virtual int FireTransfer(WebDAVTransferEventParams* e)
	{
		printf("Resource being received from server (in full text): \n========================================= \n%s\n", e->Text);
		return 0;
	}

};

void prompt(char* writeStr, const char* message) {
	printf("%s", message);
	scanf("%s", writeStr, LINE_LEN);
}


int main(int argc, char* argv[])
{
	MyWebDAV webdav;
	char command[LINE_LEN];               // user's command
	char argument[LINE_LEN];           // arguments to the user's command
	char buffer[LINE_LEN];
	int ret_code;

	if (argc < 2) {

		fprintf(stderr, "usage: webdav username password\n");
		fprintf(stderr, "username  the username to login\n");
		fprintf(stderr, "password  the password to login\n");
		fprintf(stderr, "\r\nExample: webdav username password\n");
		exit(1);
	}
	printf("******************************************************************************\n");
	printf("* This demo shows how to use the WebDAV component to make directories, move  *\n");
	printf("* resources, delete resource, and to get resources.                          *\n");
	printf("******************************************************************************\n\n");



	while (true) {
		prompt(argument, "Choose an option:\n1) Make Directory\n2) Move Resource\n3) Get Resource\n4) Delete Resource\n5) Put Resource\nQ) Quit.\n> ");

		if (!strcmp(argument, "1")) { // Make Directory
			webdav.Reset();
			webdav.SetUser(argv[argc - 2]);
			webdav.SetPassword(argv[argc - 1]);
			char server[LINE_LEN];
			char dir[LINE_LEN];
			prompt(server, "Name server where you wish to create a directory (ex. http://localhost:443): ");
			prompt(dir, "Name directory to create (ex. directory/folder): ");
			sprintf(buffer, "%s/%s", server, dir);
			ret_code = webdav.MakeDirectory(buffer);

		}
		else if (!strcmp(argument, "2")) {  // Move Resource
			webdav.Reset();
			webdav.SetUser(argv[argc - 2]);
			webdav.SetPassword(argv[argc - 1]);
			char server[LINE_LEN];
			char src[LINE_LEN];
			char dest[LINE_LEN];
			char path[LINE_LEN] = "/\0";
			prompt(server, "Name server (ex. http://localhost:443): ");
			prompt(src, "Name source of the resource (ex. myoldfolder/myfile.txt): ");
			prompt(dest, "Name destination of the resource (ex. mynewfolder/myfile.txt): ");
			strcat(path, dest);
			sprintf(buffer, "%s/%s", server, src);
			ret_code = webdav.MoveResource(buffer, dest);
		}
		else if (!strcmp(argument, "3")) { // Get Resource
			webdav.Reset();
			webdav.SetUser(argv[argc - 2]);
			webdav.SetPassword(argv[argc - 1]);
			char get[LINE_LEN];
			prompt(get, "Name URI of resource you wish to get: ");
			ret_code = webdav.GetResource(get);

		}
		else if (!strcmp(argument, "4")) { // Delete Resource
			webdav.Reset();
			webdav.SetUser(argv[argc - 2]);
			webdav.SetPassword(argv[argc - 1]);
			char del[LINE_LEN];
			prompt(del, "Name URI of resource you wish to delete: ");
			ret_code = webdav.DeleteResource(del);

		}
		else if (!strcmp(argument, "5")) { // Put Resource on server, can create resource on server or replace old resource
			webdav.Reset();
			webdav.SetUser(argv[argc - 2]);
			webdav.SetPassword(argv[argc - 1]);
			char file[LINE_LEN];
			char put[LINE_LEN];
			prompt(file, "Name path of file you wish to put on server: ");
			webdav.SetLocalFile(file);
			prompt(put, "Name URI of resource you wish to put on server: ");
			ret_code = webdav.PutResource(put);

		}
		else if (!strcmp(argument, "q") || !strcmp(argument, "Q")) {
			exit(0);
		}
		else if (!strcmp(argument, "")) {
			// Do nothing
		}
		else {
			fprintf(stderr, "Bad command / Not implemented in demo.");
		} // end of command checking
		if (ret_code)    // Got an error.
		{
			printf("\nError: %d", ret_code);
			if (webdav.GetLastError())
			{
				printf(" \"%s\"\n", webdav.GetLastError());
			}
		}
		ret_code = 0; //flush out error

	}
}

File: IPWorks Samples/WebDAV Client/python/webdav.py
# 
# IPWorks 2024 Python Edition - Sample Project
# 
# This sample project demonstrates the usage of IPWorks in a 
# simple, straightforward way. It is not intended to be a complete 
# application. Error handling and other checks are simplified for clarity.
# 
# www.nsoftware.com/ipworks
# 
# This code is subject to the terms and conditions specified in the 
# corresponding product license agreement which outlines the authorized 
# usage and restrictions.
# 

import sys
import string
from ipworks import *

input = sys.hexversion < 0x03000000 and raw_input or input


def ensureArg(args, prompt, index):
    if len(args) <= index:
        while len(args) <= index:
            args.append(None)
        args[index] = input(prompt)
    elif args[index] is None:
        args[index] = input(prompt)


webdav = WebDAV()

def fireSSLServerAuthentication(e):
    print("\nServer provided the following fingerprint:\n%s\n"%e.fingerprint)
    cntue = input("Would you like to continue [y/n]: ")
    if cntue == "Y" or cntue == "y":
      e.accept = True

	
def fireConnected(e):
	print("Server connected\n")

def fireConnectionStatus(e):	
    print("Status Code %d: %s" %(e.status_code, e.description))
	
def fireDisconnected(e):
	print("Server disconnected\n")
	
def fireTransfer(e):
	print("Resource being received from server (in full text): \n========================================= \n%s\n" % e.text)

print("******************************************************************************")
print("* This demo shows how to use the WebDAV component to make directories, move  *")
print("* resources, delete resource, and to get resources.                          *")
print("******************************************************************************\n")


if len(sys.argv) != 3:
    print("usage: python webdav_client.py username password")
    print("Options")
    print("  ")
    print("username  the username to login")
    print("password  the password to login")
    print("\r\nExample: webdav username password")
    sys.exit()
else:
    try:
        webdav.on_ssl_server_authentication = fireSSLServerAuthentication
        webdav.on_connected = fireConnected
        webdav.on_connection_status = fireConnectionStatus
        webdav.on_disconnected = fireDisconnected
        webdav.on_transfer = fireTransfer
        
        while True:
            command = input("Choose an option:\n1) Make Directory\n2) Move Resource\n3) Get Resource\n4) Delete Resource\n5) Put Resource\nQ) Quit.\n> ")
            if command == "1":
                webdav.reset()
                webdav.set_user(sys.argv[1])
                webdav.set_password(sys.argv[2])
                server = input("Name server where you wish to create a directory (ex. http://localhost:443): ")
                dir = input("Name directory to create (ex. directory/folder): ")
                webdav.make_directory(server + "/" + dir)
            elif command == "2":
                webdav.reset()
                webdav.set_user(sys.argv[1])
                webdav.set_password(sys.argv[2])
                server = input("Name server (ex. http://localhost:443): ")
                src = input("Name source of the resource (ex. myoldfolder/myfile.txt): ")
                dest = input("Name destination of the resource (ex. mynewfolder/myfile.txt): ")
                webdav.move_resource(server + "/" + src, "/" + dest)
            elif command == "3":
                webdav.reset()
                webdav.set_user(sys.argv[1])
                webdav.set_password(sys.argv[2])
                file = input("Name URI of resource you wish to get: ")
                webdav.get_resource(file)
            elif command == "4":
                webdav.reset()
                webdav.set_user(sys.argv[1])
                webdav.set_password(sys.argv[2])
                file = input("Name URI of resource you wish to delete: ")
                webdav.delete_resource(file)
            elif command == "5":
                webdav.reset()
                webdav.set_user(sys.argv[1])
                webdav.set_password(sys.argv[2])
                file = input("Name path of file you wish to put on server: ")
                webdav.set_local_file(file)
                put = input("Name URI of resource you wish to put on server: ")
                webdav.put_resource(put)
            elif command == "q" or command == "Q":
                sys.exit()
        
    except IPWorksError as e:
        print(e)



